tools/ion-bench/src/ion/reading.h

```diff
@@ -0,0 +1,200 @@
+            this->_context = {
+               MAX_DECIMAL_DIGITS,   /* max digits */
+               DEC_MAX_MATH,         /* max exponent */
+               -DEC_MAX_MATH,        /* min exponent */
+               DEC_ROUND_HALF_EVEN,  /* rounding mode */
+               DEC_Errors,           /* trap conditions */
+               0,                    /* status flags */
+               0                     /* apply exponent clamp? */
+            };
+            options.decimal_context = &(this->_context);
+            hREADER reader = nullptr;
+
+            auto err = ion_reader_open_buffer(&reader, buffer, bufsize, &options);
+            if (err == IERR_OK) {
+               _reader = reader;
+            } else {
+               printf("ERROR\n");
+            }
+         }
+
+         ~BufferReader() {
+            ion_reader_close(_reader);
+         }
+
+         ION_TYPE current_type() const {
+            return _current;
+         }
+
+         std::optional<std::string> field_name() const {
+            ION_STRING ionstr = {0};
+            ion_reader_get_field_name(_reader, &ionstr);
+            if (ionstr.length > 0) {
+               std::string value((char *)ionstr.value, ionstr.length);
+               return std::make_optional(value);
+            }
+            return std::nullopt;
+         }
+
+         iERR next() {
+            auto err = ion_reader_next(_reader, &_current);
+            return err;
+         }
+
+         bool in_struct() {
+            BOOL in = FALSE;
+            auto err = ion_reader_is_in_struct(_reader, &in);
+            if (err != IERR_OK)
+               printf("ERROR calling is_in_struct: %d\n", err);
+            return in == TRUE;
+         }
+
+         iERR step_in() {
+            return ion_reader_step_in(_reader);
+         }
+
+         iERR step_out() {
+            return ion_reader_step_out(_reader);
+         }
+
+         int depth() const {
+            int depth = 0;
+            ion_reader_get_depth(_reader, &depth);
+            return depth;
+         }
+
+         bool is_null() const {
+            BOOL is_null = FALSE;
+            auto err = ion_reader_is_null(_reader, &is_null);
+            if (err != IERR_OK)
+               printf("ERROR is_null: %d", err);
+            return (is_null == TRUE);
+         }
+
+         std::vector<std::string> get_annotations() {
+            std::vector<std::string> annot;
+            BOOL has_anns = FALSE;
+            SIZE count = 0;
+
+            ion_reader_has_any_annotations(_reader, &has_anns);
+            if (has_anns == TRUE) {
+               ion_reader_get_annotation_count(_reader, &count);
+               ION_STRING *syms = new ION_STRING[count];
+               ion_reader_get_annotations(_reader, syms, count, &count);
+               for (int i=0; i < count; i++) {
+                  annot.push_back(std::string((char *)syms[i].value, syms[i].length));
+               }
+               delete[] syms;
```

Comment:
Should we use `ion_reader_get_an_annotation` in the loop and avoid copying into a temporary array?

Reply:
Good call. I'll follow up with a PR to change that.